packages/@n8n/task-runner-python/src/task_analyzer.py

```diff
@@ -0,0 +1,205 @@
+            # Check keyword argument (name="module")
+            elif node.keywords:
+                for keyword in node.keywords:
+                    if keyword.arg == "name" and isinstance(
+                        keyword.value, ast.Constant
+                    ):
+                        module_arg = keyword.value.value
+                        break
+
+            if module_arg and isinstance(module_arg, str):
+                self._validate_import(module_arg, node.lineno)
+            else:
+                self._add_violation(node.lineno, ERROR_DYNAMIC_IMPORTLIB)
+
+        self.generic_visit(node)
+
+    # ========== Validation ==========
+
+    def _validate_import(self, module_path: str, lineno: int) -> None:
+        """Validate that a module import is allowed based on allowlists. Relative imports are disallowed."""
+
+        if module_path.startswith("."):
+            self._add_violation(lineno, ERROR_RELATIVE_IMPORT)
+            return
+
+        module_name = module_path.split(".")[0]  # e.g., os.path -> os
+
+        if module_name in self.checked_modules:
+            return
+
+        self.checked_modules.add(module_name)
+
+        is_stdlib = module_name in sys.stdlib_module_names
+        is_external = not is_stdlib
+
+        if is_stdlib and (self._stdlib_allow_all or module_name in self.stdlib_allow):
+            return
+
+        if is_external and (
+            self._external_allow_all or module_name in self.external_allow
+        ):
+            return
+
+        error, allowed_str = (
+            (ERROR_STDLIB_DISALLOWED, self._stdlib_allowed_str)
+            if is_stdlib
+            else (ERROR_EXTERNAL_DISALLOWED, self._external_allowed_str)
+        )
+
+        self._add_violation(
+            lineno, error.format(module=module_path, allowed=allowed_str)
+        )
+
+    def _format_allowed(self, allow_set: Set[str]) -> str:
+        return ", ".join(sorted(allow_set)) if allow_set else "none"
+
+    def _add_violation(self, lineno: int, message: str) -> None:
+        self.violations.append(f"Line {lineno}: {message}")
+
+
+class TaskAnalyzer:
+    _cache: ValidationCache = OrderedDict()
+
+    def __init__(self, stdlib_allow: Set[str], external_allow: Set[str]):
+        self._stdlib_allow = stdlib_allow
+        self._external_allow = external_allow
+        self._allowlists = (
+            tuple(sorted(stdlib_allow)),
+            tuple(sorted(external_allow)),
+        )
+        self._allow_all = "*" in stdlib_allow and "*" in external_allow
+
+    def validate(self, code: str) -> None:
+        if self._allow_all:
+            return
+
+        cache_key = self._to_cache_key(code)
+        cached_violations = self._cache.get(cache_key)
+        cache_hit = cached_violations is not None
+
+        if cache_hit and len(cached_violations) == 0:
+            return
+
+        if cache_hit and len(cached_violations) > 0:
+            self._raise_security_error(cached_violations)
```

Comment:
Should we move the cache entry to the end so more often touched entries have higher probability to stay in the cache?